kubelet/rkt: merge environments instead of overriding.

yifan-gu opened the pull request.

yifan-gu commented:
This PR changes the on-disk systemd service file name. Before this, all pod with the same UID has the same service file name `foo.service`. This causes at least 2 problems when pod foo restarts for some reason: 
- 1. We lose the old pod's service file, thus the info about this dead pod.
- 2. We will fail to run the new service file because of the naming collision. (Not sure why systemd.reload doesn't work here).

The concern is we will leave moreservice files on disk, but this shouldn't be an issue once we are able to gc them(as well as containers, images).
/cc @dchen1107 @yujuhong

yifan-gu commented:
Related to #8951

dchen1107 commented:
Above lgtm is not from me. The only thing I did is just open this pr. :-(

thockin commented:
Dawn, you should open a ticket with GitHub to see where that came from.

@brendandburns maybe we should halt bots until we understand this?

Commit:
kubelet/rkt: merge environments instead of overriding.

Merge the environments in container spec and in the image
manifest. Also allow empty app in image manifest.

dchen1107 commented:
Since I am OOT since tomorrow, reassign it to @yujuhong Thanks!

yifan-gu commented:
@dchen1107  Travel safe :)

yujuhong commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -190,11 +191,31 @@ func (r *runtime) runCommand(args ...string) ([]string, error) {
+// makePodServiceFileName constructs the unit file name for a pod using
+// its pod UID and rkt UUID. We use the rkt UUID here because it's not
+// clear how to update and reload the service file once the service remains
+// after exited.
```
If I understand correctly, you simply want a unique ID for the file name. Why would you need to update the service file?

yifan-gu replied:
@yujuhong Good question. So the situation here is 1 kube pod can map to 2 rkt pods (imagine the first rkt pod crashed, and we are about to start the second one).
For some reason(I don't recall now, but seems to be related to `RemainAfterExit`) it's not updated. Let's just hold one and let me just try `ReloadOrRestartUnit`.

yujuhong replied:
Sure. SGTM.

yujuhong commented:
LGTM with one question

Commit:
kubelet/rkt: include the pod restart count in service file.

Also remove the service file when the pod is killed.

yifan-gu commented:
@yujuhong PR updated. Removed some unnecessary functions. Now one pod will only have one service file to map with. If the service file's content is updated, we will invoke 'systemctl daemon-reload' (which equals to dbus.Conn.Reload()) to reload the service file and then restart the service.
Upon deletion, the service file will be removed as well.

yujuhong commented:
LGTM.
